[Issue #3712][python-client] exposing InitialPosition management in the ConsumerConfiguration.

lelabo-marc opened the pull request:
Fixes #3712

### Motivation
PR #3567 introduced the SubscriptionInitialPosition option in ConsumerConfiguration in the CPP client but did not exposed any methods to be able to do it with the Python client.

This PR aims to expose the code introduced in the previous PR to allows Python user to choose the initial position of the consumer in python. 

### Modifications

- Implemented a boost object to expose InitialPosition Enum.
- Added to boost ConsumerConfiguration object the getter/setter of the InitialPosition attribute.
- Added a initial_position parameter to Client.subscribe in order to modify the ConsumerConfiguration instance created.

### Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

  - Anything that affects deployment: (don't know but it should not)

### Documentation

This PR introduce a new feature for the Python client.
I added a small description of how to use it by modifying the pulsar/__init__.py as I think the sources are used to generate the documentation.

lelabo-marc commented:
Hi @merlimat, can you please take a look ?
Did my best for the tests but did not achieve to run the docker scripts.
However, I compilled the source and runned a few tests on a remote standalone without crossing any errors.
I don't think they should be many side effects as I just exposed code from the CPP client using boost and did not changed the default behaviour of the consumer.

merlimat commented on pulsar-client-cpp/python/pulsar/__init__.py:
```diff
@@ -536,6 +537,10 @@ def my_listener(consumer, message):
           Periods of seconds for consumer to auto discover match topics.
+        * `initial_position`:
+          Set the initial position of a consumer  when subscribing to the topic.
+          It could be either: `InitialPosition.InitialPositionEarliest` or `InitialPosition.InitialPositionEarliest`.
```
`Earliest` or `Latest`

merlimat commented on pulsar-client-cpp/python/pulsar/__init__.py:
```diff
@@ -472,7 +472,8 @@ def subscribe(self, topic, subscription_name,
                   properties=None,
-                  pattern_auto_discovery_period=60
+                  pattern_auto_discovery_period=60,
+                  initial_position=None
```
The default here is `Latest` anywyay, so we could set it to that instead of None

merlimat commented:
@lelabo-marc You can run the test by spinning up a test standalone cluster:  `pulsar-client-cpp/pulsar-test-service-start.sh` 

Then `cd pulsar-client-cpp/python` and `python pulsar_test.py`

lelabo-marc commented:
Followed your recommendation and implemented a new test. I also corrected the doc and shortened the name of the enum parameters.

However, 3 tests are currently failing but they seems to be not related to the current PR. The python tests are successful.

merlimat commented:
run integration tests

merlimat commented:
run java8 tests

merlimat commented:
run java8 tests
run integration tests

sijie commented:
run java8 tests

sijie commented:
run java8 tests

sijie commented:
run cpp tests

merlimat commented:
Merged in 2.3.1 at 
c078e71
